README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rNodal

The goal of rNodal is to provide nodal analysis for oil and gas wells.

## Installation

You can install rNodal from github with:

``` r
# install.packages("devtools")
devtools::install_github("f0nzie/rNodal")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
## basic example code
```

## Using `zFactor`

`zFactor` is a R package. Calling from CRAN.

We will use zFactor for gas compressibility calculations.

``` r
# use the new library zFactor
library(zFactor)

z.HallYarborough(pres.pr = 4.5, temp.pr = 1.4)
#> [1] 0.7373812
z.DranchukAbuKassem(pres.pr = 4.5, temp.pr = 1.4)
#> [1] 0.7396345
z.BeggsBrill(pres.pr = 4.5, temp.pr = 1.4)
#> [1] 0.7343367
z.Ann10(pres.pr = 4.5, temp.pr = 1.4)
#> [1] 0.736032
z.Papp(pres.pr = 4.5, temp.pr = 1.4)
#> [1] 0.7299354
```

## How rNodal works

Start by looking at the examples in the vignettes. We will use in this
example `VLP Brown - Example C13`.

This is example C.13 in the Kermit Brown book.

### Input well data

We enter the well data with the function `setWellInput`:

``` 
input.example.C13 <- setWellInput(field.name = "HAGBR.MOD",
                                    well.name = "Brown_C13", 
                                    depth.wh = 0, depth.bh = 2670, 
                                    diam.in = 1.995, 
                                    GLR = 500, liq.rt = 1000, wcut = 0.6, 
                                    thp = 500, tht = 120, bht = 150, 
                                    API = 22, gas.sg = 0.65, 
                                    wat.sg = 1.07, if.tens = 30)
                                    
                                    
```

The field name and well name are used for archival purposes.

### Enter the parameters of the VLP model

The parameters of the model consist of:

`vlp.model`: the correlation or mechanistical model

`segments`: the number of segments to split the well

`tol`: the tolerance of the delta-P iterations

``` 
well.model <- setVLPmodel(vlp.model = "hagbr.mod", 
                           segments = 11, 
                                tol = 0.000001)
                                
                                
                                
```

### Run the model

To run the model is necessary to provide:

`well.input`: all the well parameters as entered in the first step

`well.model`: the VLP model as entered in the second step

    runVLP(well.input = input.example.C13, well.model))

### Results

The results are given in the form of a dataframe where the rows
represent the number of segment plus one and the columns are the
calculations or variables.

<img src="man/figures/README-results_df.jpg" width="800px" />


NEWS.md
## 20190202
* 

## 20170729 0.0.2.9003
* push to remote
* remove unused line in travis


## 20170714 0.0.2.9002
* remove two unit tests in test_nodal_status.R. They were too explicit to the source folders and making the pkg crash.
* running coverage results. Had to remove two test functions in test_LL_paths.R
        > covr::package_coverage()
        rNodal Coverage: 52.18%
        R/all_generics.R: 0.00%
        R/class-VLP.R: 0.00%
        R/Fancher-Brown.R: 0.00%
        R/heat_transfer.R: 0.00%
        R/interpolation.R: 0.00%
        R/Poettmann-Carpenter.R: 0.00%
        R/zzz.R: 0.00%
        R/friction.R: 1.79%
        R/settings.R: 7.41%
        R/class-well_deviation_survey.R: 30.67%
        R/Duns-Ros.R: 35.56%
        R/utils.R: 42.70%
        R/gas_correlations.R: 57.66%
        R/hdf5.R: 73.33%
        R/Hagedorn-Brown.R: 74.70%
        R/oil_correlations.R: 78.05%
        R/VLP.R: 94.79%
        

## 20170710 0.0.2.9001
* add Hmisc to Imports to use list.tree in vignette C13
* use pres. average instead of pres at the end
* change name of test dataset to brown_oilwell01.rda
* add function temp.fluid to calculate fluid temperaturem to heat_transger.R
* add VLP class to class-VLP.R. But not used yet
* add notebook to test VLP class
* add new script all_generics.R
* calculate the mass flow rate w = m * q
* move mass.rate object from Hagedorn-Brown.R to VLP.R
* copy example from heat_transfer notebook to vignette C13
* add U, oil.cp, gas.cp, wat.cp to well input
* add calculation of Cp average
* add new function get_well_parameters that combines the well input and the basic calculations
* add functions shift and na.zero to calculate the angle of a deviation survey.
* finding angle from MD, TVD working
* add diam.organize notebook on "build deviation survey with angles". add sections.ft object because causes vignette C13 to crash
* Merge branch 'mass.rate_to_well_params' into feature/0.0.2.9001
* Merge branch 'feature/0.0.2.9001' into develop

## 20170709 0.0.2.9000
* rename VLP scripts to author names separated by dashes.
* move functions in dunsros.R to own script.
* move any gas fucntion to gas_correlations.R
* rename FLUIDPROPS.R to oil_correlations.R
* move functions in zfactor.R to gas_correlations.R
* remove data.R. move its functions to rNodal-package.r
* no Rd for .saveSlot function

## 20170709 0.0.2
* document few functions
* ignore folder notebooks


## 20170709 0.0.1.9003
* use Hall-Yarborough from zFactor
* remove old HY built-in in this package
* new datasets in testthat regenerated with new values of z with zFactor
* add more tests to test_zfactor.R
* use zFactor::HallYarborough directly in the tests supplying sequences of Ppr and Tpr that are in a reasonable range of oil wells.
* using zFactor::HallYarborough at low and high Ppr, Tpr
* create a rda file for test check zFactor with returning lists of z, pres.pr, temp.pr and temp.r


## 20170709 0.0.1.9002
* add links to h5 files in vignettes
* h5 temp file is the same for all vignettes VLP runs
* VLP tables are added to h5 in temp file unless R session is restarted
* add word test to all test files
* add new test for all VLP runs that were in vignettes. 
* each VLP test now has its own rda file for test check
* add verbose to functions VLPControl and hdf5 creation function
* rename surf.params to well.params

## 20170707 0.0.1.9001
* change source of notebooks to ../R
* re-running vignettes
* remove html and R from vignettes
* rename one vignette
* add title to index vignette


## 20170706 0.0.1.9000
* all tests running ok
* move folders from inst to notebooks
* improve strategy


## 20170706 0.0.1
* keeping 9012 as a base
* changes to z return values were discarded; causing problems

## 20170611 9012
* fix two tests in test_nodal_status.R because of a change of folder in rNodal.

## 20170611 9011
* add new file STRATEGY.md

## 20170611 9010
* changes in z factor
* new equation for z.
* strategy to extract single values instead of list

## 20170605 9008
* FIX problem with interpolation.R. It was renamed but this causes conflicts 
with existing file in the repository because the old file INTERPOLATION.R is not
removed or renamed. Fixed by moving `interpolation.R` to another folder and then
bringing it back.
* Next time, instead of renaming a file try `Save As`.
* Tag 0.0.0.9007, 9008 deleted.


## 20170605 9006
* complete low level tests adding test_that
* tried to use subfolders under testthat bu was not succesful
* rename tests to be able to group them later. Low level tests now have a tag LL.
* change name and references of file INTERPOLATION.R to lowercase.
* rename data file used in tests. Start with `data`.


## 20170604 0.0.0.9003
* running OK
* rename get_extdataDir
* add getDefaultDataContainerName
* new test is_checking_package
* remove warnings in settings.R


## 20170604 0.0.0.9002
* Implemented a way to capture if we are in build package mode using a new function is_checking_package.
* added a bunch of small tests for testing paths.
* change logical function to something like this: is_xxx_yyy(). Example: 
    * is_checking_package
    * is_saved_session
* can copy a HDF5 data container in "local" mode. Similar to "project". 
* can check if the user has already a data comtainer so it is not overwritten,


## 20170531 `rPetroleum` package
* Found problem with rhdf5 which did not allow to build the package. Problem is with lack of declaration as S3 method. Found solution in forums. Issue is known. Fix is to manually edit NAMESPACE and add an `import(rhdf5)`. 

R/VLP.R
#' @include utils.R settings.R  gas_correlations.R
#' @include friction.R oil_correlations.R interpolation.R
NULL


#  ########################
#' Globals
#'
PRES.ATM     = 14.7      # psia
TEMP.STD     = 60        # degrees Farenheit
TEMP.RANKINE = 460       # to convert to absolute temperature


# setProjectEnvironment()

saveToProjectEnv("PRES.ATM", 14.7)
saveToProjectEnv("TEMP.STD", 60)
saveToProjectEnv("TEMP.RANKINE", 460)



#' Build a deviation survey dataframe given MD and TVD as text table
#'
#' The table should be two columns of text
#'
#' @param well_as_string well deviation survey as two columns text
#' @export
set_deviation_survey <- function(well_as_string) {
    # read string text to dataframe
    well_table <- utils::read.table(header = TRUE, text = well_as_string)
    well_table
}





#' Set the high level model parameters for VLP calculations
#'
#' Prepare high level model inputs such as type of model, number of tubing segments,
#' tolerance for iterations, initial value of dp/dz putting them as a list.
#' Default VLP model is Hagendorn-Brown-Guo
#'
#' @param vlp.model     the name of the VLP model or correlation
#' @param segments      number of segments for the tubing string
#' @param tol           tolerance for error during interations
#' @param dp.dz.ini     initial gradient
#' @export
setVLPmodel <- function( vlp.model = "hagbr.guo",  # name of the VLP correlation
                         segments = 29,            # table rows = segments + 1
                         tol = 0.0001,             # tolerance in dp.dz calc
                         dp.dz.ini = 0.002         # initial value for dp.dz
                         ) {

    named.list(vlp.model,
               segments,
               tol,
               dp.dz.ini
               )
}



#' Set the most common well inputs
#'
#' Prepare all the well inputs to be entered into the VLP model by putting all
#' the inputs in a list.
#'
#' @param field.name field name. Usually comprises several wells
#' @param well.name  well name
#' @param depth.wh  depth at wellhead
#' @param depth.bh  depth at bottomhole of well                    feet
#' @param diam.in   well diameter                                  inches
#' @param GLR       gas liquid ratio
#' @param liq.rt    liquid rate
#' @param wcut      watercut
#' @param thp       tubing head pressure
#' @param tht       tubing head temperature
#' @param bht       bottomhole temperature
#' @param API       gravity of oil
#' @param gas.sg    specific gravity of gas
#' @param wat.sg    specific gravity of water
#' @param oil.visc  oil viscosity
#' @param ed        relative rougness
#' @param if.tens   interfacial tension between ...
#' @param salinity  water salinity
#' @param oil.cp    heat capacity of oil
#' @param gas.cp    heat capacity of gas
#' @param wat.cp    heat capacity of water
#' @param U         overall heat transfer coefficient
#' @param angle     angle of the well to the horizontal
#' @param geotherm geothermal data at two TVD points
#' @rdname setWellInput-VLP
#' @export
setWellInput <- function( field.name = "HAGBR.GUO",
                            well.name = "ExampleGuo",
                            depth.wh  = 0,               # depth at wellhead
                            depth.bh  = 9700,            # depth at bottomhole
                            diam.in   = 1.995,
                            GLR       = 75,
                            liq.rt    = 758,
                            wcut      = 0.10,
                            thp       = 200,       # tubing head pressure
                            tht       = 80,        # tubing head temperature
                            bht       = 180,       # bottomhole temperature
                            API       = 40,
                            gas.sg    = 0.7,
                            wat.sg    = 1.05,
                            oil.visc  = 5,
                            ed        = 0.0006,      # relative roughness
                            if.tens   = 30,
                            salinity  = 0,
                            U         = 8.0,         # heat transfer coefficient
                            oil.cp    = 0.53,        # heat capacity
                            gas.cp    = 0.5,
                            wat.cp    = 1.0,
                            angle     = pi/2,
                          geotherm    = NULL
                        ) {

    # well input parameters
    out <- named.list(      field.name,
                            well.name,
                            depth.wh, tht,
                            depth.bh, bht,
                            diam.in, ed,
                            thp, liq.rt, wcut,
                            API, oil.visc,
                            gas.sg, GLR,
                            wat.sg,
                            salinity,
                            if.tens,
                            U,
                            oil.cp, gas.cp, wat.cp,
                            angle,
                            geotherm
    )
    return(out)
}


# load default values for well input
Gwell.input <- setWellInput()


#' Perform basic calculations based on the well inputs
#'
#' @param well.input well input data as a list
#' @export
getBasicCalcs <- function(well.input) {
    with(as.list(well.input), {

        # convert text table of geothermal gradient to dataframe
        if (!is.null(geotherm)) {  # do not calculate if no geot data provided
            geotherm_df <- read.table(header = TRUE, text = geotherm)
            bht <- tail(geotherm_df, 1)[["temp"]]
            tht <- head(geotherm_df, 1)[["temp"]]
            depth.bh <- tail(geotherm_df, 1)[["TVD"]] - head(geotherm_df, 1)[["TVD"]]
        } else {
            geotherm_df <- data.frame()
            warning("No geothermal table entered")
        }

        # calculate temperature gradient
        temp.grad <- (bht - tht) / depth.bh

        # convert tubing diameter to ft
        diam <- diam.in /12
        diam.ft <- diam.in / 12

        # calculate area in ft^2
        area <- pi / 4 * diam^2

        # calculate specific gravity of oil from API
        oil.sg <- 141.5 / (131.5 + API)

        # oil and water fractions
        wat.fraction <- wcut
        oil.fraction <- 1 - wat.fraction
        WOR          <- wat.fraction / oil.fraction

        # calculate oil, gas and water rate at standard conditions
        # Options of names to use: oil.srt, gas.srt, wat.srt
        #                          OIL.RT, GAS.RT, WAT.RT
        #                          oil.Srt, gas.Srt, wat.Srt
        oil.rt <- liq.rt * oil.fraction
        gas.rt <- liq.rt * GLR
        wat.rt <- liq.rt * wat.fraction

        # GOR
        GOR = (oil.rt + wat.rt) / oil.rt * GLR

        # total mass per STB = mass oil + mass water + mass gas. C42.3
        mass.total <- oil.sg * 350 * (1 / (1+WOR)) +
            wat.sg * 350 * (WOR / (1+WOR)) +
            0.0764 * GLR * gas.sg

        # TODO: calculate fluid properties at P, T conditions

        # 4. calculate the mass flow rate w = m * q
        mass.rt  <-  mass.total * liq.rt
        mass.rate <- mass.rt

        # heat capacity
        cp.avg <- (oil.cp + gas.cp + wat.cp) /3

        # calculated
        out.calc <- named.list( temp.grad,
                                diam, area, diam.ft,         # added diam.ft
                                oil.sg,
                                oil.fraction, wat.fraction, WOR,
                                oil.rt, gas.rt, wat.rt,
                                mass.total,
                                GOR,
                                mass.rt, mass.rate,
                                cp.avg
        )
        return(out.calc)
    })
}


#' Get the well input together with the basic calculations
#'
#' @param well.input the well input
#' @export
get_well_parameters <- function(well.input) {
    # perform basic calculations on the well input
    basic.calcs <-  getBasicCalcs(well.input)
    well.parameters <- c(well.input, basic.calcs)
    well.parameters
}



#' Run the VLP calculations based on well inputs and model parameters
#'
#' Calls VLPcontrol() to run the subsusrface calculations
#'
#' @param well.input        well input data as a list
#' @param model.parameters  well model parameters as a list
#' @export
runVLP <- function(well.input, model.parameters) {

    # perform basic calculations on the well input
    basic_calcs <-  getBasicCalcs(well.input)
    # well_parameters <- c(well.input, basic_calcs) # join input and calc output

    # pass the well parameters and model parameters
    vlp.output <- VLPcontrol(well.input, basic_calcs, model.parameters)

    return(tibble::as_tibble(vlp.output))                 # return dataframe
}



#' VLP calculation algorithm
#'
#' VLP control marching algorithm to calculate bottomhole pressure given the
#' wellhead pressure and temperature. Fluid properties are calculated at each
#' of the segment depths.
#'
#' @param well_input  unprocessed list of well inputs            list
#' @param basic_calcs  basic calculations based on the well inputs  list
#' @param model_parameters   model characteristics. Hagedorn-Brown, Duns-Ros,
#'                           Fancher-Brown, etc. Also tolerances and boundaries
#' @param verbose prevent printing messages. Default is FALSE
#'
#' @export
VLPcontrol <- function(well_input, basic_calcs, model_parameters,
                       verbose = FALSE) {
    # called by runVLP()
    with(as.list(c(well_input, basic_calcs, model_parameters)),
    {
        if (verbose) cat("VLP control for well model:", vlp.model, "\n")

        # load the VLP function provided in model_parameters
        vlp.function = loadVLP(vlp.model)

        # Calculate the well segments and depths
        # Depth points have to be greater than segments to allocate the zero or
        # initial depth value. Consider that in R for the "length.out" parameter.
        # Mind that in R index starts at 1 not 0 like C or Python
        depths <- seq.int(from = depth.wh, to = depth.bh, length.out = segments+1)
        n      <- length(depths)   # which is the same as # rows in the dataframe
        depth.top <- depths[1]                # take the the first depth

        dp.dz     <- dp.dz.ini                # 1st approximation of the gradient
        p0        <- thp                      # the initial pressure
        t0        <- tht                      # initial temperature
        dt.dz     <- temp.grad                # temperature gradient at inlet

        # initialize vectors of list type to store row calculations
        segment_row_vector <- vector("list", n)
        iter_row_vector    <- vector("list")

        cum_iter <- 1                      # counter for all iterations
        for (i in seq_len(n)) {            # n is the number of depths = # rows
            # make the previous depth the top but do this only for the 1st row
            # otherwise, use the previous depth
            if (i == 1) depth.prev = depth.top else depth.prev = depths[i-1]
            dL  <- depths[i] - depth.prev       # calculate dL
            # prepare p1, t1 for iteration
            p1  <- p0 + dp.dz * dL              # calculate outlet pressure
            t1  <- t0 + dt.dz * dL              # calculate outlet temperature
            eps <-  1                           # initial value for epsilon
            iter_dpdz <- 1                      # AE: absolute error
            # here we start iterating for the pressure gradient
            # until it is within the specified tolerance
            while (eps > tol) {           # loop until AE greater than tolerance
                p.avg <- (p0 + p1) / 2    # try with an initial pressure
                t.avg <- (t0 + t1) / 2
                # calculate pressure losses using the selected VLP correlation
                # passing the current pressure, temperature and other parameters
                corr  <- vlp.function(pres = p.avg, temp = t.avg,
                                      well_input, basic_calcs)

                dp.dz <- corr$dp.dz       # extract dp/dz or pressure gradient
                z     <- corr$z
                # calculate new pressure
                p.calc <- p0 - (-dp.dz) * dL # negative, we are going down
                eps    <- abs( (p1 - p.calc) / p.calc )  # absolute error
                # build iteration row-vector
                iter_row_vector[[cum_iter]] <- list(
                    segment = i,
                    iter_dpdz = as.integer(iter_dpdz),
                    epsilon = eps,
                    p0 = p0, p1 = p1, p.calc = p.calc,
                    p.avg = p.avg, delta.P = abs(p1-p.calc),
                    depth = depths[i], dL = dL,
                    dp.dz = dp.dz, temp = t1
                    )
                if (eps >= tol) p1 = p.calc   # if error too big, iterate again
                iter_dpdz <- iter_dpdz + 1    # with a a new p1
                cum_iter <- cum_iter + 1      # number of total iterations
            } # end-while
            # at the end of the while-loop, we obtain the average pressure
            # then, build a row-vector out of: depth, dL, temperature, pressure,
            # segment, correlation results
            segment_row_vector[[i]] <- c(
                i = i,             # row number
                depth = depths[i], # depth
                dL = dL,           # length of pipe increment
                temp = t1,         # current temperature
                pres = p1,         # current pressure at depth
                p_avg = p.avg,
                t_avg = t.avg,
                segment = i-1,     # segment number
                corr               # correlation results
                )
            p0 = p.calc   # assign p1 to the inlet pressure of new segment, p0
            t0 = t1       # do the same with the temperature
    } # end-for
        iter.tbl <- data.table::rbindlist(iter_row_vector) # row-list to dataframe
    # convert row-vector to a dataframe
    segment_tbl <- data.table::rbindlist(segment_row_vector)
    return(segment_tbl)                      # results in segments table
    }) # end with
}


runVLPdefaults <- function() {
    # FIX: error inside loop with `tol`
    well.inputs      <- setWellInput()
    model.parameters <- setVLPmodel()

    runVLP(well.input = well.inputs, model.parameters = model.parameters)
    #return(well.inputs)
    #return(c(well.inputs, model.parameters))
}


#' Load only the source necessary for model or correlation
#' Note: it doesn't unload the functions sourced yet
#' @param model   the model name       string
#'
loadVLP <- function(model) {
    modelU <- toupper(model)
    # if (grepl("HAGBR", modelU))    source("HAGBR.R")
    # if (grepl("DUNSROS", modelU))  source("DUNSROS.R")
    # if (grepl("FANBR", modelU))    source("FANBR.R")

    if (model == "hagbr.guo")      return(hagbr.guo)
    if (model == "hagbr.dummy")    return(hagbr.dummy)
    if (model == "hagbr.mod")      return(hagbr.mod)       # <-
    if (model == "dunsros.0")      return(dunsros.0)
    if (model == "fanbr.fanbr")    return(fanbr.fanbr)
}












